test-data/unit/check-partially-defined.test

```diff
@@ -367,3 +367,48 @@ def f() -> None:
     d = a
     d = b
 [builtins fixtures/tuple.pyi]
+
+[case testUseBeforeDef]
+# flags: --enable-error-code use-before-def
+
+def f0() -> None:
+    x = y  # E: Name "y" is used before definition
+    y: int = 1
+
+def f1() -> None:
+    if int():
+        x = 0
+    else:
+        y = x  # E: Name "x" is used before definition
+
+def f2() -> None:
+    x = 1
+    if int():
+        x = 0
+    else:
+        y = x  # No error.
+
+def f3() -> None:
+    if int():
+        pass
+    else:
+        # No use-before-def error.
+        y = z  # E: Name "z" is not defined
+
+    def inner2() -> None:
+        z = 0
+
+def f4() -> None:
+    if int():
+        pass
+    else:
+        y = z  # E: Name "z" is used before definition
+    z: int = 2
+
+def f5() -> None:
+    if int():
+        pass
+    else:
+        y = z  # E: Name "z" is used before definition
+        x = z  # No error here -- we shouldn't report the error twice.
```

Comment:
I think we should actually report both cases. Imagine you have
```python
def some_long_func() -> None:
    x = z
    # 50 lines later
    y = z
```
In this case someone will need to fix first error, then run mypy again to get the second one and so on.

Reply:
I think it may be a bit cleaner to just report one error but let's report all and then see what real-world issues we see with it.